File: Aed's_II/Verde/Java/Games.java
package Verde.Java;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Games
{
// valores (variaveis) do arquivo csv
    public int                          id; 
    public int              estimatedOwners;
    public int                 achievements;
    public int              metacriticScore;
    public double                     price;
    public double                 userScore;
    public String                      name;
    public String               releaseDate;
    public String [ ]    supportedLanguages;
    public String [ ]            publishers;
    public String [ ]            developers;
    public String [ ]            categories;
    public String [ ]                genres;
    public String [ ]                  tags;

    public Games( 
    // construtor basico para criacao de valores setados
     int id, int estimatedOwners, int achievements,
     int metacriticScore, double price, double userScore, String name, String releaseDate, 
     String [ ] supportedLanguages, String [ ] publishers, String [ ] developers, 
     String [ ] categories, String [ ] genres, String [ ] tags ) 
    {
        this.id                     =               id;
        this.estimatedOwners        =   estimatedOwners;
        this.achievements           =      achievements;
        this.metacriticScore        =   metacriticScore;
        this.price                  =            price;
        this.userScore              =         userScore;
        this.name                   =              name;
        this.releaseDate            =       releaseDate;
        this.supportedLanguages     =   supportedLanguages;
        this.publishers             =        publishers;
        this.developers             =        developers;
        this.categories             =        categories;
        this.genres                 =            genres;
        this.tags                   =              tags;

        ERRO( );    // normalizar campos
    }

    //---------------------------||---------------------------\\
    
    public void ERRO ( ) 
    {
    //---------------------------||---------------------------\\
        // Corrigir data
        if ( releaseDate != null && releaseDate.trim().length() > 0 )
        {
            String [ ] partes = releaseDate.split("/");
            String novaData = "";
            int i = 0;

            while ( i < 3 )
            {
                if ( i < partes.length && partes[i].trim().length() > 0 )
                { novaData = novaData + partes[i]; }
                else
                { novaData = novaData + "01"; }

                if ( i < 2 )
                { novaData = novaData + "/"; }

                i = i + 1;
            }

            releaseDate = novaData;
        }

    //---------------------------||---------------------------\\
        if ( estimatedOwners < 0 )
        { estimatedOwners = 0; }

    //---------------------------||---------------------------\\
        if ( price < 0 )
        { price = 0.0; }

    //---------------------------||---------------------------\\
        if ( metacriticScore <= 0 )
        { metacriticScore = -1; }

    //---------------------------||---------------------------\\
        if ( userScore <= 0.0 )
        { userScore = -1.0; }

    //---------------------------||---------------------------\\
        if ( achievements < 0 )
        { achievements = 0; }

    //---------------------------||---------------------------\\
    }

    public static class CSVreader 
    {
        public static List<Games> readCSV(String filePath) 
        { 
            List <Games> gamesList = new ArrayList<Games>(); 

            try
            {
                BufferedReader br = new BufferedReader(new FileReader(filePath));
                String line = br.readLine();  // descartar cabeçalho

                line = br.readLine();

                while ( line != null )
                {
                    if ( line.trim().length() > 0 )
                    {
                        String [ ] fields = line.split(",");

                        if ( fields.length >= 14 )
                        {
                            int id                       = parseIntSafe (fields [0]);
                            String name                  = fields [ 1 ];
                            String releaseDate           = fields [ 2 ];
                            int estimatedOwners          = parseIntSafe(fields[3]);
                            double price                 = parseDoubleSafe(fields[4]);
                            String[] supportedLanguages  = toArray(fields[5]);
                            int metacriticScore          = parseIntSafe(fields[6]);
                            double userScore             = parseDoubleSafe(fields[7]);
                            int achievements             = parseIntSafe(fields[8]);
                            String[] publishers          = toArray(fields[9]);
                            String[] developers          = toArray(fields[10]);
                            String[] categories          = toArray(fields[11]);
                            String[] genres              = toArray(fields[12]);
                            String[] tags                = toArray(fields[13]);

                            Games game = new Games (
                                id, estimatedOwners, achievements,
                                metacriticScore, price, userScore,
                                name, releaseDate,
                                supportedLanguages, publishers,
                                developers, categories,
                                genres, tags
                            );

                            gamesList.add(game);
                        }
                    }
                    line = br.readLine();
                }

                br.close();

            }
            catch (IOException e)
            {
                e.printStackTrace();
            }

            return gamesList;
        }

        //---------------------------||---------------------------\\
        private static int parseIntSafe (String s)
        {
            int value = 0;
            if ( s != null && s.trim().length() > 0 )
            {
                try
                { value = Integer.parseInt(s.trim().replaceAll("[^0-9]", "")); }
                catch (Exception e)
                { value = 0; }
            }
            return value;
        }

        //---------------------------||---------------------------\\
        private static double parseDoubleSafe (String s)
        {
            double value = 0.0;
            if ( s != null && s.trim().length() > 0 )
            {
                try
                { value = Double.parseDouble(s.trim().replaceAll("[^0-9.]", "")); }
                catch (Exception e)
                { value = 0.0; }
            }
            return value;
        }

        //---------------------------||---------------------------\\
        private static String [ ] toArray (String s)
        {
            String [ ] arr = new String [0];

            if ( s != null && s.trim().length() > 0 )
            {
                s = s.replace("[", "");
                s = s.replace("]", "");

                arr = s.split(",");

                int i = 0;
                while ( i < arr.length )
                {
                    arr[i] = arr[i].trim();
                    i = i + 1;
                }
            }
            return arr;
        }
        //---------------------------||---------------------------\\
    }

    //---------------------------||---------------------------\\
    @Override
    public String toString ( )
    {
        return "Game [id=" + id + ", name=" + name +
               ", releaseDate=" + releaseDate +
               ", owners=" + estimatedOwners +
               ", price=" + price +
               ", meta=" + metacriticScore +
               ", userScore=" + userScore + "]";
    }

    //---------------------------||---------------------------\\
    public static void main ( String [ ] args ) 
    {
        String filePath = "tmp/games.csv";

        // Ler os jogos
        List <Games> gamesList = Games.CSVreader.readCSV ( filePath );

        // Mostrar quantos foram carregados
       // System.out.println ( "Total de jogos carregados: " + gamesList.size() );

        // Exibir todos
        //System.out.println ( "\nLista completa de jogos:\n" );

        int i = 0;
        while ( i < gamesList.size() )
        {
            System.out.println ( gamesList.get(i) );
            i = i + 1;
        }
    }

}

File: Aed's_II/Verde/C/Games.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_STR      200
#define MAX_LIST     1000
#define MAX_ARRAY    50

//---------------------------||---------------------------\\

typedef struct Games
{
    int        id;
    int        estimatedOwners;
    int        achievements;
    int        metacriticScore;
    double     price;
    double     userScore;
    char       name[MAX_STR];
    char       releaseDate[MAX_STR];
    char       supportedLanguages[MAX_ARRAY][MAX_STR];
    char       publishers[MAX_ARRAY][MAX_STR];
    char       developers[MAX_ARRAY][MAX_STR];
    char       categories[MAX_ARRAY][MAX_STR];
    char       genres[MAX_ARRAY][MAX_STR];
    char       tags[MAX_ARRAY][MAX_STR];

    int        nSupportedLanguages;
    int        nPublishers;
    int        nDevelopers;
    int        nCategories;
    int        nGenres;
    int        nTags;

} Games;

//---------------------------||---------------------------\\
// funções auxiliares

int parseIntSafe ( const char *s )
{
    int value = 0;
    int i = 0;
    char tmp[MAX_STR];
    int j = 0;

    if ( s != NULL && strlen(s) > 0 )
    {
        while ( s[i] != '\0' )
        {
            if ( isdigit( (unsigned char) s[i] ) )
            {
                tmp[j] = s[i];
                j = j + 1;
            }
            i = i + 1;
        }
        tmp[j] = '\0';

        if ( j > 0 )
        { value = atoi(tmp); }
    }

    return value;
}

//---------------------------||---------------------------\\

double parseDoubleSafe ( const char *s )
{
    double value = 0.0;
    if ( s != NULL && strlen(s) > 0 )
    {
        value = atof(s);
    }
    return value;
}

//---------------------------||---------------------------\\

int toArray ( char array[MAX_ARRAY][MAX_STR], const char *s )
{
    int count = 0;
    char copy[MAX_STR * 2];
    char *token;

    if ( s != NULL && strlen(s) > 0 )
    {
        strcpy(copy, s);
        // remover colchetes
        char *p = copy;
        int i = 0;
        while ( copy[i] != '\0' )
        {
            if ( copy[i] == '[' || copy[i] == ']' )
            { copy[i] = ' '; }
            i = i + 1;
        }

        token = strtok(copy, ",");
        while ( token != NULL && count < MAX_ARRAY )
        {
            while ( *token == ' ' ) token++;
            strncpy(array[count], token, MAX_STR - 1);
            array[count][MAX_STR - 1] = '\0';

            count = count + 1;
            token = strtok(NULL, ",");
        }
    }

    return count;
}

//---------------------------||---------------------------\\

void normalizar ( Games *g )
{
    // Corrigir data
    if ( strlen(g->releaseDate) > 0 )
    {
        char tmp[MAX_STR];
        strcpy(tmp, g->releaseDate);

        char *partes[3];
        int n = 0;
        char *token = strtok(tmp, "/");

        while ( token != NULL && n < 3 )
        {
            partes[n] = token;
            n = n + 1;
            token = strtok(NULL, "/");
        }

        char novaData[MAX_STR] = "";
        int i = 0;

        while ( i < 3 )
        {
            if ( i < n && strlen(partes[i]) > 0 )
            {
                strcat(novaData, partes[i]);
            }
            else
            {
                strcat(novaData, "01");
            }

            if ( i < 2 )
            { strcat(novaData, "/"); }

            i = i + 1;
        }

        strcpy(g->releaseDate, novaData);
    }

    if ( g->estimatedOwners < 0 )
    { g->estimatedOwners = 0; }

    if ( g->price < 0 )
    { g->price = 0.0; }

    if ( g->metacriticScore <= 0 )
    { g->metacriticScore = -1; }

    if ( g->userScore <= 0.0 )
    { g->userScore = -1.0; }

    if ( g->achievements < 0 )
    { g->achievements = 0; }
}

//---------------------------||---------------------------\\

int readCSV ( const char *filePath, Games gamesList[MAX_LIST] )
{
    FILE *fp = fopen(filePath, "r");
    char line[2048];
    int count = 0;

    if ( fp == NULL )
    {
        printf("Erro ao abrir arquivo!\n");
        return 0;
    }

    // descartar cabeçalho
    fgets(line, sizeof(line), fp);

    while ( fgets(line, sizeof(line), fp) != NULL && count < MAX_LIST )
    {
        // remover \n
        line[strcspn(line, "\n")] = '\0';

        if ( strlen(line) > 0 )
        {
            char *fields[14];
            int i = 0;

            char *token = strtok(line, ",");
            while ( token != NULL && i < 14 )
            {
                fields[i] = token;
                i = i + 1;
                token = strtok(NULL, ",");
            }

            if ( i >= 14 )
            {
                Games g;

                g.id                  = parseIntSafe(fields[0]);
                strncpy(g.name, fields[1], MAX_STR - 1);
                g.name[MAX_STR - 1] = '\0';

                strncpy(g.releaseDate, fields[2], MAX_STR - 1);
                g.releaseDate[MAX_STR - 1] = '\0';

                g.estimatedOwners     = parseIntSafe(fields[3]);
                g.price               = parseDoubleSafe(fields[4]);
                g.nSupportedLanguages = toArray(g.supportedLanguages, fields[5]);
                g.metacriticScore     = parseIntSafe(fields[6]);
                g.userScore           = parseDoubleSafe(fields[7]);
                g.achievements        = parseIntSafe(fields[8]);
                g.nPublishers         = toArray(g.publishers, fields[9]);
                g.nDevelopers         = toArray(g.developers, fields[10]);
                g.nCategories         = toArray(g.categories, fields[11]);
                g.nGenres             = toArray(g.genres, fields[12]);
                g.nTags               = toArray(g.tags, fields[13]);

                normalizar(&g);

                gamesList[count] = g;
                count = count + 1;
            }
        }
    }

    fclose(fp);
    return count;
}

//---------------------------||---------------------------\\

void printGame ( const Games *g )
{
    printf("Game [id=%d, name=%s, releaseDate=%s, owners=%d, price=%.2f, meta=%d, userScore=%.2f]\n",
           g->id, g->name, g->releaseDate, g->estimatedOwners, g->price, g->metacriticScore, g->userScore );
}

//---------------------------||---------------------------\\

int main ( )
{
    Games gamesList[MAX_LIST];
    int total = 0;
    int i = 0;

    total = readCSV("tmp/games.csv", gamesList);

    //printf("Total de jogos carregados: %d\n", total);
   // printf("\nLista completa de jogos:\n\n");

    i = 0;
    while ( i < total )
    {
        printGame(&gamesList[i]);
        i = i + 1;
    }

    return 0;
}
